--- cyrus-sasl-2.1.17/configure.in.orig	2003-12-03 23:15:12.000000000 +0100
+++ cyrus-sasl-2.1.17/configure.in	2003-12-03 23:24:56.368097984 +0100
@@ -695,9 +695,7 @@
     AC_ERROR([--enable-sql chosen but neither Postgres nor MySQL found])
 fi
 
-if test "$ac_cv_can_build_shared" = yes; then
 	AC_DEFINE(DO_DLOPEN,[],[Should we build a shared plugin (via dlopen) library?])
-fi
 
 AC_SUBST(SASL_MECHS)
 AC_SUBST(SASL_STATIC_OBJS)
