--- cyrus-sasl-1.5.27/configure.in.orig	Sun Aug 12 17:41:40 2001
+++ cyrus-sasl-1.5.27/configure.in	Sun Aug 12 17:41:57 2001
@@ -169,8 +169,8 @@
 	dnl We want to attempt to use -lndbm if we can, just in case
 	dnl there's some version of it installed and overriding libc
 	AC_CHECK_HEADER(ndbm.h,
-			AC_CHECK_LIB(ndbm, dbm_open, SASL_DB_LIB="-lndbm",
-				AC_CHECK_FUNC(dbm_open,,dblib="no")),
+			[AC_CHECK_LIB(ndbm, dbm_open, SASL_DB_LIB="-lndbm",
+				[AC_CHECK_FUNC(dbm_open,,dblib="no")])],
 				dblib="no")
 	;;
   auto_detect)
