--- cyrus-sasl-2.1.23/configure.in.orig	2011-08-03 10:26:16.234418985 +0200
+++ cyrus-sasl-2.1.23/configure.in	2011-08-03 10:37:21.689418838 +0200
@@ -86,9 +86,6 @@
 AC_PROG_INSTALL
 CMU_C___ATTRIBUTE__
 
-dnl check for -R, etc. switch
-CMU_GUESS_RUNPATH_SWITCH
-
 dnl xxx compatibility
 AC_ARG_WITH(staticsasl)
 if test "$with_staticsasl" = yes; then
@@ -265,10 +265,6 @@
 	with_pam=$withval,
 	with_pam=yes)
 if test "$with_pam" != no; then
-  if test -d $with_pam; then
-    CPPFLAGS="$CPPFLAGS -I${with_pam}/include"
-    LDFLAGS="$LDFLAGS -L${with_pam}/lib"
-  fi
   AC_CHECK_HEADERS(security/pam_appl.h pam/pam_appl.h)
   cmu_save_LIBS="$LIBS"
   AC_CHECK_FUNC(pam_start, :, 
@@ -461,8 +461,7 @@
 		;;
 	*)
 		if test -d $with_opie; then
-		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include -I${with_opie}/include/security"
-		  LDFLAGS="${LDFLAGS} -L${with_opie}/lib"
+		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include/security"
 		else
 		  with_opie="no"
 		fi
@@ -692,16 +691,6 @@
     no) true;;
     notfound) AC_WARN([MySQL Library not found]); true;;
     *)
-     if test -d ${with_mysql}/lib/mysql; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/lib/mysql, LIB_MYSQL)
-     elif test -d ${with_mysql}/mysql/lib; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/mysql/lib, LIB_MYSQL)
-     elif test -d ${with_mysql}/lib; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/lib, LIB_MYSQL)
-     else
-	CMU_ADD_LIBPATH_TO(${with_mysql}, LIB_MYSQL)
-     fi
-
      SASL_MECHS="$SASL_MECHS libmysql.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_MYSQL_DIR=$LIB_MYSQL
@@ -758,16 +747,6 @@
     no) true;;
     notfound) AC_WARN([PostgreSQL Library not found]); true;;
     *)
-     if test -d ${with_pgsql}/lib/pgsql; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/lib/pgsql, LIB_PGSQL)
-     elif test -d ${with_pgsql}/pgsql/lib; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/pgsql/lib, LIB_PGSQL)
-     elif test -d ${with_pgsql}/lib; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/lib, LIB_PGSQL)
-     else
-	CMU_ADD_LIBPATH_TO(${with_pgsql}, LIB_PGSQL)
-     fi
-
      SASL_MECHS="$SASL_MECHS libpgsql.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_PGSQL_DIR=$LIB_PGSQL
@@ -825,12 +804,6 @@
     no) true;;
     notfound) AC_WARN([SQLite Library not found]); true;;
     *)
-     if test -d ${with_sqlite}/lib; then
-         LIB_SQLITE="-L${with_sqlite}/lib -R${with_sqlite}/lib"
-     else
-         LIB_SQLITE="-L${with_sqlite} -R${with_sqlite}"
-     fi
-
      SASL_MECHS="$SASL_MECHS libsqlite.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_SQLITE_DIR=$LIB_SQLITE
@@ -872,11 +845,6 @@
     save_CPPFLAGS=$CPPFLAGS
     save_LDFLAGS=$LDFLAGS
 
-    if test -d $with_ldap; then
-        CPPFLAGS="${CPPFLAGS} -I${with_ldap}/include"
-        CMU_ADD_LIBPATH(${with_ldap}/lib)
-    fi
-
     AC_CHECK_HEADERS(ldap.h lber.h)
 
     if test $ac_cv_header_ldap_h = yes -a $ac_cv_header_lber_h = yes; then
--- cyrus-sasl-2.1.23/cmulocal/ax_path_bdb.m4~	2005-01-06 21:24:52.000000000 +0100
+++ cyrus-sasl-2.1.23/cmulocal/ax_path_bdb.m4	2011-08-03 10:49:59.141418659 +0200
@@ -89,8 +89,6 @@
        # Check for library
        AX_PATH_BDB_NO_OPTIONS([$1], [ENVONLY], [
          ax_path_bdb_ok=yes
-         BDB_CPPFLAGS="-I$ax_path_bdb_INC"
-         BDB_LDFLAGS="-L$ax_path_bdb_LIB"
        ])
      else
        AC_MSG_RESULT([no])
@@ -240,8 +238,6 @@
       if test "$ax_compare_version" = "true" ; then
         ax_path_bdb_no_options_ok=yes
         BDB_LIBS="-ldb"
-        BDB_CPPFLAGS="-I$ax_path_bdb_path_find_highest_DIR/include"
-        BDB_LDFLAGS="-L$ax_path_bdb_path_find_highest_DIR/lib"
         BDB_VERSION="$ax_path_bdb_path_find_highest_VERSION"
       fi
     fi
@@ -337,7 +333,6 @@
     LIBS="$LIBS -ldb"
 
     ax_path_bdb_path_get_version_save_LDFLAGS="$LDFLAGS"
-    LDFLAGS="-L$1/lib $LDFLAGS"
 
     # Compile and run a program that compares the version defined in
     # the header file with a version defined in the library function
--- cyrus-sasl-2.1.23/saslauthd/configure.in~	2011-08-03 11:35:35.854770408 +0200
+++ cyrus-sasl-2.1.23/saslauthd/configure.in	2011-08-03 11:37:56.047770375 +0200
@@ -28,7 +28,6 @@
 
 dnl Checks for build foo
 CMU_C___ATTRIBUTE__
-CMU_GUESS_RUNPATH_SWITCH
 
 dnl Checks for libraries.
 CMU_SOCKETS
@@ -92,10 +91,6 @@
 	with_pam=$withval,
 	with_pam=yes)
 if test "$with_pam" != no; then
-  if test -d $with_pam; then
-    CPPFLAGS="$CPPFLAGS -I${with_pam}/include"
-    LDFLAGS="$LDFLAGS -L${with_pam}/lib"
-  fi
   cmu_save_LIBS="$LIBS"
   AC_CHECK_LIB(pam, pam_start, [
 	  AC_CHECK_HEADER(security/pam_appl.h,,
@@ -132,11 +127,6 @@
 	with_ldap=no)
 AC_MSG_RESULT($with_ldap)
 
-if test -d $with_ldap; then
-    CPPFLAGS="$CPPFLAGS -I${with_ldap}/include"
-    CMU_ADD_LIBPATH(${with_ldap}/lib)
-fi
-
 LDAP_LIBS=""
 if test "$with_ldap" != no; then
   AC_CHECK_LIB(ldap, ldap_initialize, [ AC_DEFINE(HAVE_LDAP,[],[Support for LDAP?])
