--- cyrus-sasl-2.1.18/sasldb/Makefile.am.wiget	2004-03-10 16:51:36.000000000 +0100
+++ cyrus-sasl-2.1.18/sasldb/Makefile.am	2004-03-11 17:20:16.214506300 +0100
@@ -51,7 +51,6 @@
 EXTRA_DIST = NTMakefile
 
 noinst_LTLIBRARIES = libsasldb.la
-noinst_LIBRARIES = libsasldb.a
 
 libsasldb_la_SOURCES = allockey.c sasldb.h
 EXTRA_libsasldb_la_SOURCES = $(extra_common_sources)
@@ -63,8 +62,3 @@
 # Prevent make dist stupidity
 libsasldb_a_SOURCES =
 EXTRA_libsasldb_a_SOURCES =
-
-libsasldb.a: libsasldb.la $(LOCAL_SASL_DB_BACKEND_STATIC)
-	$(AR) cru .libs/$@ $(LOCAL_SASL_DB_BACKEND_STATIC)
-
-
