--- cyrus-sasl-2.1.10/lib/Makefile.am.orig	Fri Dec  6 17:24:00 2002
+++ cyrus-sasl-2.1.10/lib/Makefile.am	Thu Jan  9 20:49:51 2003
@@ -66,6 +66,9 @@
 plugin_common.lo: plugin_common.o
 	rm -f plugin_common.lo
 	ln -s $(top_builddir)/plugins/plugin_common.lo plugin_common.lo
+	if [ -f $(top_builddir)/plugins/.libs/plugin_common.o ]; then \
+		ln -s ../$(top_builddir)/plugins/.libs/plugin_common.o .libs/plugin_common.o ; \
+	fi
 
 plugin_common.o:
 	rm -f plugin_common.o
