--- cyrus-sasl-1.5.28/saslauthd/Makefile.am~	Thu Oct 14 20:26:01 2004
+++ cyrus-sasl-1.5.28/saslauthd/Makefile.am	Thu Oct 14 20:26:47 2004
@@ -1,7 +1,7 @@
 sbin_PROGRAMS	= saslauthd
 
 saslauthd_SOURCES = saslauthd.c mechanisms.c globals.h mechanisms.h auth_dce.c auth_dce.h auth_getpwent.c auth_getpwent.h auth_krb5.c auth_krb5.h auth_krb4.c auth_krb4.h auth_pam.c auth_pam.h auth_rimap.c auth_rimap.h auth_shadow.c auth_shadow.h auth_sia.c auth_sia.h
-saslauthd_LDADD	= @SASL_KRB_LIB@ @GSSAPIBASE_LIBS@ @GSSAPI_LIBS@ @PLAIN_LIBS@ @LIB_SOCKET@
+saslauthd_LDADD	= @SASL_KRB_LIB@ @GSSAPIBASE_LIBS@ @GSSAPI_LIBS@ @PLAIN_LIBS@ @LIB_SOCKET@ @LIB_PAM@ @LIB_CRYPT@
 
 EXTRA_DIST	= saslauthd.8
 INCLUDES	= -I$(top_srcdir)/include
