--- cyrus-sasl-2.1.27/lib/Makefile.am.orig	2018-11-24 09:23:55.021657715 +0100
+++ cyrus-sasl-2.1.27/lib/Makefile.am	2018-11-24 09:24:00.648324318 +0100
@@ -96,25 +96,6 @@
 install-exec-hook:
 endif
 
-libsasl2.a: libsasl2.la $(SASL_STATIC_OBJS)
-	@echo adding static plugins and dependencies
-	$(AR) cru .libs/$@ $(SASL_STATIC_OBJS)
-	@for i in ./libsasl2.la ../common/libplugin_common.la ../sasldb/libsasldb.la ../plugins/lib*.la; do \
-	if test ! -f $$i; then continue; fi; . $$i; \
-	for j in $$dependency_libs foo; do \
-	case $$j in foo) ;; \
-	-L*) for k in $$depdirs foo; do \
-	    if test $$k = $$j; then break; fi; done; \
-	  if test $$k = foo; then depdirs="$$depdirs $$j"; fi ;; \
-	-l*) for k in $$deplibs foo; do \
-	    if test $$k = $$j; then break; fi; done; \
-	  if test $$k = foo; then deplibs="$$deplibs $$j"; fi ;; \
-	esac; done; dependency_libs=""; done; \
-	sed -e "/^dependency_libs=/s%=.*%='$${depdirs}$${deplibs}'%" \
-	libsasl2.la >TMP.$$ && mv TMP.$$ libsasl2.la
-	rm -f $@
-	ln -s .libs/$@ $@
-
 $(SASL_STATIC_SRCS): linksrcs
 
 linksrcs:
