--- saslauthd/auth_httpform.c~	2009-02-16 22:59:05.000000000 +0200
+++ saslauthd/auth_httpform.c	2009-02-16 23:00:16.765691380 +0200
@@ -176,7 +176,7 @@
         return NULL;
 
     while (inidx < length) {
-        char in = string[inidx];
+        unsigned char in = string[inidx];
         if (!(in >= 'a' && in <= 'z') &&
             !(in >= 'A' && in <= 'Z') &&
             !(in >= '0' && in <= '9') &&
