diff -urN cyrus-sasl-2.1.20.org/configure.in cyrus-sasl-2.1.20/configure.in
--- cyrus-sasl-2.1.20.org/configure.in	2004-10-24 22:05:13.000000000 +0200
+++ cyrus-sasl-2.1.20/configure.in	2005-04-04 23:12:57.615789008 +0200
@@ -584,9 +584,7 @@
 AC_MSG_CHECKING(SQL)
 if test "$sql" != no; then
   AC_MSG_RESULT(enabled)
-  SASL_MECHS="$SASL_MECHS libsql.la"
   if test "$enable_static" = yes; then
-    SASL_STATIC_OBJS="$SASL_STATIC_OBJS ../plugins/sql.o"
     AC_DEFINE(STATIC_SQL,[],[Link SQL plugin staticly])
   fi
 else
@@ -633,6 +631,8 @@
 	CMU_ADD_LIBPATH_TO(${with_mysql}, LIB_MYSQL)
      fi
 
+     SASL_MECHS="$SASL_MECHS libmysql.la"
+     SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_MYSQL_DIR=$LIB_MYSQL
      LIB_MYSQL="$LIB_MYSQL -lmysqlclient"
 
@@ -697,6 +697,8 @@
 	CMU_ADD_LIBPATH_TO(${with_pgsql}, LIB_PGSQL)
      fi
 
+     SASL_MECHS="$SASL_MECHS libpgsql.la"
+     SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_PGSQL_DIR=$LIB_PGSQL
      LIB_PGSQL="$LIB_PGSQL -lpq"
 
@@ -758,6 +760,8 @@
          LIB_SQLITE="-L${with_sqlite} -R${with_sqlite}"
      fi
 
+     SASL_MECHS="$SASL_MECHS libsqlite.la"
+     SASL_STATIC_OBJS="$SASL_STATIC_OBJS"
      LIB_SQLITE_DIR=$LIB_SQLITE
      LIB_SQLITE="$LIB_SQLITE -lsqlite"
 
diff -urN cyrus-sasl-2.1.20.org/plugins/Makefile.am cyrus-sasl-2.1.20/plugins/Makefile.am
--- cyrus-sasl-2.1.20.org/plugins/Makefile.am	2004-10-24 22:05:20.000000000 +0200
+++ cyrus-sasl-2.1.20/plugins/Makefile.am	2005-04-04 23:22:49.454815800 +0200
@@ -77,7 +77,7 @@
 sasl_LTLIBRARIES = @SASL_MECHS@
 EXTRA_LTLIBRARIES = libplain.la libanonymous.la libkerberos4.la libcrammd5.la \
 	libgssapiv2.la libdigestmd5.la liblogin.la libsrp.la libotp.la \
-	libntlm.la libsasldb.la libsql.la
+	libntlm.la libsasldb.la libmysql.la libpgsql.la libsqlite.la
 
 libplain_la_SOURCES = plain.c plain_init.c $(common_sources)
 libplain_la_LDFLAGS = -version-info $(plain_version)
@@ -135,17 +135,41 @@
 libsasldb_la_DEPENDENCIES = $(COMPAT_OBJS)
 libsasldb_la_LIBADD = ../sasldb/libsasldb.la $(SASL_DB_LIB) $(COMPAT_OBJS)
 
-libsql_la_SOURCES = sql.c sql_init.c $(common_sources)
-libsql_la_LDFLAGS = $(LIB_MYSQL) $(LIB_PGSQL) $(LIB_SQLITE) -version-info $(sql_version)
-libsql_la_DEPENDENCIES = $(COMPAT_OBJS)
-libsql_la_LIBADD = $(COMPAT_OBJS)
-
+libmysql_la_SOURCES = mysql.c mysql_init.c $(common_sources)
+libmysql_la_LDFLAGS = $(LIB_MYSQL)  -version-info $(sql_version)
+libmysql_la_DEPENDENCIES = $(COMPAT_OBJS)
+libmysql_la_LIBADD = $(COMPAT_OBJS)
+
+libpgsql_la_SOURCES = pgsql.c pgsql_init.c $(common_sources)
+libpgsql_la_LDFLAGS = $(LIB_PGSQL) -version-info $(sql_version)
+libpgsql_la_DEPENDENCIES = $(COMPAT_OBJS)
+libpgsql_la_LIBADD = $(COMPAT_OBJS)
+
+libsqlite_la_SOURCES = sqlite.c sqlite_init.c $(common_sources)
+libsqlite_la_LDFLAGS = $(LIB_SQLITE) -version-info $(sql_version)
+libsqlite_la_DEPENDENCIES = $(COMPAT_OBJS)
+libsqlite_la_LIBADD = $(COMPAT_OBJS)
+
+mysql.c:
+	sed -e 's#HAVE_PGSQL#HAVE_NO_PGSQL#g' -e 's#HAVE_SQLITE#HAVE_NO_SQLITE#g' \
+		-e 's#sql_auxprop_plug_init#mysql_auxprop_plug_init#g' \
+		sql.c > mysql.c
+
+pgsql.c:
+	sed -e 's#HAVE_MYSQL#HAVE_NO_MYSQL#g' -e 's#HAVE_SQLITE#HAVE_NO_SQLITE#g' \
+		-e 's#sql_auxprop_plug_init#pgsql_auxprop_plug_init#g' \
+		sql.c > pgsql.c
+
+sqlite.c:
+	sed -e 's#HAVE_MYSQL#HAVE_NO_MYSQL#g' -e 's#HAVE_PGSQL#HAVE_NO_PGSQL#g' \
+		-e 's#sql_auxprop_plug_init#sqlite_auxprop_plug_init#g' \
+		sql.c > sqlite.c
 
 # Instructions for making the _init files
 
 init_src=anonymous_init.c crammd5_init.c digestmd5_init.c gssapiv2_init.c \
 kerberos4_init.c login_init.c plain_init.c srp_init.c otp_init.c ntlm_init.c \
-sasldb_init.c sql_init.c
+sasldb_init.c mysql_init.c pgsql_init.c sqlite_init.c
 
 
 CLEANFILES=$(init_src)
diff -urN cyrus-sasl-2.1.20.org/plugins/makeinit.sh cyrus-sasl-2.1.20/plugins/makeinit.sh
--- cyrus-sasl-2.1.20.org/plugins/makeinit.sh	2003-07-15 19:38:59.000000000 +0200
+++ cyrus-sasl-2.1.20/plugins/makeinit.sh	2005-04-04 23:00:03.871750928 +0200
@@ -45,7 +45,7 @@
 " > ${mech}_init.c
 done
 
-for mech in sasldb sql ; do
+for mech in sasldb mysql pgsql sqlite ; do
 
 echo "
 #include <config.h>
