--- cyrus-sasl-2.1.27/configure.ac.orig	2018-11-24 13:32:32.178154053 +0100
+++ cyrus-sasl-2.1.27/configure.ac	2018-11-24 17:29:45.091324872 +0100
@@ -104,9 +104,6 @@
 AC_PROG_INSTALL
 AC_USE_SYSTEM_EXTENSIONS
 
-dnl check for -R, etc. switch
-CMU_GUESS_RUNPATH_SWITCH
-
 dnl xxx compatibility
 AC_ARG_WITH(staticsasl)
 if test "$with_staticsasl" = yes; then
@@ -479,8 +476,7 @@
 		;;
 	*)
 		if test -d $with_opie; then
-		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include -I${with_opie}/include/security"
-		  LDFLAGS="${LDFLAGS} -L${with_opie}/lib"
+		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include/security"
 		else
 		  with_opie="no"
 		fi
@@ -800,16 +796,6 @@
      LDFLAGS=$save_LDFLAGS
      ;;
     *)
-     if test -d ${with_mysql}/lib/mysql; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/lib/mysql, LIB_MYSQL)
-     elif test -d ${with_mysql}/mysql/lib; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/mysql/lib, LIB_MYSQL)
-     elif test -d ${with_mysql}/lib; then
-	CMU_ADD_LIBPATH_TO(${with_mysql}/lib, LIB_MYSQL)
-     else
-	CMU_ADD_LIBPATH_TO(${with_mysql}, LIB_MYSQL)
-     fi
-
      SASL_MECHS="$SASL_MECHS libmysql.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS mysql.o"
      LIB_MYSQL_DIR=$LIB_MYSQL
@@ -878,16 +864,6 @@
      LDFLAGS=$save_LDFLAGS
      ;;
     *)
-     if test -d ${with_pgsql}/lib/pgsql; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/lib/pgsql, LIB_PGSQL)
-     elif test -d ${with_pgsql}/pgsql/lib; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/pgsql/lib, LIB_PGSQL)
-     elif test -d ${with_pgsql}/lib; then
-	CMU_ADD_LIBPATH_TO(${with_pgsql}/lib, LIB_PGSQL)
-     else
-	CMU_ADD_LIBPATH_TO(${with_pgsql}, LIB_PGSQL)
-     fi
-
      SASL_MECHS="$SASL_MECHS libpgsql.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS pgsql.o"
      LIB_PGSQL_DIR=$LIB_PGSQL
@@ -947,11 +923,6 @@
     no) true;;
     notfound) AC_WARN([SQLite Library not found]); true;;
     *)
-     if test -d ${with_sqlite}/lib; then
-         CMU_ADD_LIBPATH_TO(${with_sqlite}/lib, LIB_SQLITE)
-     else
-         CMU_ADD_LIBPATH_TO(${with_sqlite}, LIB_SQLITE)
-     fi
 
      SASL_MECHS="$SASL_MECHS libsqlite.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS sqlite.o"
@@ -1001,11 +972,6 @@
     no) true;;
     notfound) AC_WARN([SQLite3 Library not found]); true;;
     *)
-     if test -d ${with_sqlite3}/lib; then
-         CMU_ADD_LIBPATH_TO(${with_sqlite3}/lib, LIB_SQLITE3)
-     else
-         CMU_ADD_LIBPATH_TO(${with_sqlite3}, LIB_SQLITE3)
-     fi
 
      SASL_MECHS="$SASL_MECHS libsqlite3.la"
      SASL_STATIC_OBJS="$SASL_STATIC_OBJS sqlite3.o"
@@ -1048,11 +1014,6 @@
     save_CPPFLAGS=$CPPFLAGS
     save_LDFLAGS=$LDFLAGS
 
-    if test -d $with_ldap; then
-        CPPFLAGS="${CPPFLAGS} -I${with_ldap}/include"
-        CMU_ADD_LIBPATH(${with_ldap}/lib)
-    fi
-
     AC_CHECK_HEADERS(ldap.h lber.h)
 
     if test $ac_cv_header_ldap_h = yes -a $ac_cv_header_lber_h = yes; then
