--- cyrus-sasl-2.1.23/cmulocal/bsd_sockets.m4.orig	2011-08-06 20:20:49.396403059 +0200
+++ cyrus-sasl-2.1.23/cmulocal/bsd_sockets.m4	2011-08-12 19:07:57.869787230 +0200
@@ -9,14 +9,14 @@
 AC_DEFUN([CMU_SOCKETS], [
 	save_LIBS="$LIBS"
 	LIB_SOCKET=""
-	AC_CHECK_FUNC(connect, :,
+	AC_CHECK_FUNC(connect, :,[
 		AC_CHECK_LIB(nsl, gethostbyname,
 			     LIB_SOCKET="-lnsl $LIB_SOCKET")
 		AC_CHECK_LIB(socket, connect,
-			     LIB_SOCKET="-lsocket $LIB_SOCKET")
+			     LIB_SOCKET="-lsocket $LIB_SOCKET")]
 	)
 	LIBS="$LIB_SOCKET $save_LIBS"
-	AC_CHECK_FUNC(res_search, :,
+	AC_CHECK_FUNC(res_search, :,[
 		LIBS="-lresolv $LIB_SOCKET $save_LIBS"
 		AC_TRY_LINK([[
 #include <sys/types.h>
@@ -31,7 +31,7 @@
 res_search( host, C_IN, T_MX, (u_char *)&ans, sizeof(ans));
 return 0;
 ]], LIB_SOCKET="$LIB_SOCKET") dnl res_* funcs aren't used anyway
-        )
+        ])
 	LIBS="$LIB_SOCKET $save_LIBS"
 	AC_CHECK_FUNCS(dn_expand dns_lookup)
 	LIBS="$save_LIBS"
