--- cyrus-sasl-2.1.27.org/configure.ac	2004-10-24 22:05:13.000000000 +0200
+++ cyrus-sasl-2.1.27/configure.ac	2005-04-04 22:13:49.575508064 +0200
@@ -420,7 +420,7 @@
 		;;
 	*)
 		if test -d $with_opie; then
-		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include"
+		  CPPFLAGS="${CPPFLAGS} -I${with_opie}/include -I${with_opie}/include/security"
 		  LDFLAGS="${LDFLAGS} -L${with_opie}/lib"
 		else
 		  with_opie="no"
