--- cyrus-sasl-2.1.21/sasldb/Makefile.am.orig	2005-05-07 06:54:18.000000000 +0200
+++ cyrus-sasl-2.1.21/sasldb/Makefile.am	2005-05-20 22:52:50.027548768 +0200
@@ -51,7 +51,6 @@
 EXTRA_DIST = NTMakefile
 
 noinst_LTLIBRARIES = libsasldb.la
-noinst_LIBRARIES = libsasldb.a
 
 libsasldb_la_SOURCES = allockey.c sasldb.h
 EXTRA_libsasldb_la_SOURCES = $(extra_common_sources)
@@ -61,8 +60,3 @@
 # Prevent make dist stupidity
 libsasldb_a_SOURCES =
 EXTRA_libsasldb_a_SOURCES =
-
-libsasldb.a: libsasldb.la $(SASL_DB_BACKEND_STATIC)
-	$(AR) cru .libs/$@ $(SASL_DB_BACKEND_STATIC)
-
-
--- cyrus-sasl-2.1.25/lib/Makefile.am.orig	2011-09-05 16:18:10.000000000 +0200
+++ cyrus-sasl-2.1.25/lib/Makefile.am	2011-09-16 18:57:36.749138290 +0200
@@ -76,25 +76,6 @@
 install-exec-hook:
 endif
 
-libsasl2.a: libsasl2.la $(SASL_STATIC_OBJS)
-	@echo adding static plugins and dependencies
-	$(AR) cru .libs/$@ $(SASL_STATIC_OBJS)
-	@for i in ./libsasl2.la ../sasldb/libsasldb.la ../plugins/lib*.la; do \
-	if test ! -f $$i; then continue; fi; . $$i; \
-	for j in $$dependency_libs foo; do \
-	case $$j in foo) ;; \
-	-L*) for k in $$depdirs foo; do \
-	    if test $$k = $$j; then break; fi; done; \
-	  if test $$k = foo; then depdirs="$$depdirs $$j"; fi ;; \
-	-l*) for k in $$deplibs foo; do \
-	    if test $$k = $$j; then break; fi; done; \
-	  if test $$k = foo; then deplibs="$$deplibs $$j"; fi ;; \
-	esac; done; dependency_libs=""; done; \
-	sed -e "/^dependency_libs=/s%=.*%='$${depdirs}$${deplibs}'%" \
-	libsasl2.la >TMP.$$ && mv TMP.$$ libsasl2.la
-	rm -f $@
-	ln -s .libs/$@ $@
-
 $(SASL_STATIC_SRCS): linksrcs
 
 linksrcs:
