--- cyrus-sasl-2.1.27/docsrc/exts/sphinxlocal/writers/manpage.py~	2016-11-07 14:52:12.000000000 +0100
+++ cyrus-sasl-2.1.27/docsrc/exts/sphinxlocal/writers/manpage.py	2021-09-29 00:15:02.216786169 +0200
@@ -23,7 +23,6 @@
 from sphinx import addnodes
 from sphinx.locale import admonitionlabels, _
 from sphinx.util.osutil import ustrftime
-from sphinx.util.compat import docutils_version
 
 class CyrusManualPageWriter(ManualPageWriter):
 
